include/time_units.h

```diff
@@ -0,0 +1,855 @@
+ */
+
+#ifndef ZEPHYR_INCLUDE_TIME_UNITS_H_
+#define ZEPHYR_INCLUDE_TIME_UNITS_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/* Exhaustively enumerated, highly optimized time unit conversion API */
+
+#if defined(CONFIG_TIMER_READS_ITS_FREQUENCY_AT_RUNTIME)
+__syscall int z_clock_hw_cycles_per_sec_runtime_get(void);
+
+static inline int z_impl_z_clock_hw_cycles_per_sec_runtime_get(void)
+{
+	extern int z_clock_hw_cycles_per_sec;
+
+	return z_clock_hw_cycles_per_sec;
+}
+#endif /* CONFIG_TIMER_READS_ITS_FREQUENCY_AT_RUNTIME */
+
+static inline int sys_clock_hw_cycles_per_sec(void)
+{
+#if defined(CONFIG_TIMER_READS_ITS_FREQUENCY_AT_RUNTIME)
+	return z_clock_hw_cycles_per_sec_runtime_get();
+#else
+	return CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC;
+#endif
+}
+
+/* Time converter generator gadget.  Selects from one of three
+ * conversion algorithms: ones that take advantage when the
+ * frequencies are an integer ratio (in either direction), or a full
+ * precision conversion.  Clever use of extra arguments causes all the
+ * selection logic to be optimized out, and the generated code even
+ * reduces to 32 bit only if a ratio conversion is available and the
+ * result is 32 bits.
+ *
+ * This isn't intended to be used directly, instead being wrapped
+ * appropriately in a user-facing API.  The boolean arguments are:
+ *
+ *    const_hz  - The hz arguments are known to be compile-time
+ *                constants (because otherwise the modulus test would
+ *                have to be done at runtime)
+ *    result32  - The result will be truncated to 32 bits on use
+ *    round_up  - Return the ceiling of the resulting fraction
+ *    round_off - Return the nearest value to the resulting fraction
+ *                (pass both round_up/off as false to get "round_down")
+ */
+static ALWAYS_INLINE u64_t z_tmcvt(u64_t t, u32_t from_hz, u32_t to_hz,
+				   bool const_hz, bool result32,
+				   bool round_up, bool round_off)
+{
+	bool mul_ratio = const_hz &&
+		(to_hz > from_hz) && ((to_hz % from_hz) == 0);
+	bool div_ratio = const_hz &&
+		(from_hz > to_hz) && ((from_hz % to_hz) == 0);
+
+	if (from_hz == to_hz) {
+		return result32 ? ((u32_t)t) : t;
+	}
+
+	if (!mul_ratio) {
+		u32_t rdivisor = div_ratio ? (from_hz / to_hz) : from_hz;
+
+		if (round_up) {
+			t += rdivisor - 1;
+		} else if (round_off) {
+			t += rdivisor / 2;
+		}
+	}
+
+	/* Select (at build time!) between three different expressions for
+	 * the same mathematical relationship, each expressed with and
+	 * without truncation to 32 bits (I couldn't find a way to make
+	 * the compiler correctly guess at the 32 bit result otherwise).
+	 */
+	if (div_ratio) {
+		if (result32) {
+			return ((u32_t)t) / (from_hz / to_hz);
+		} else {
+			return t / (from_hz / to_hz);
+		}
+	} else if (mul_ratio) {
+		if (result32) {
+			return ((u32_t)t) * (to_hz / from_hz);
+		} else {
+			return t * (to_hz / from_hz);
+		}
+	} else {
+		if (result32) {
+			return (((u32_t)t) * to_hz) / from_hz;
```

Comment:
I'm skeptical of the correctness of this, because it truncates the numerator to 32-bits before doing a division.  If `t * to_hz` exceeds 2^32-1 the result appears to be wrong.

I ran the code in this context:
```
  u64_t t = 1U << 31;
  u32_t from_hz = 13;
  u32_t to_hz = 7;

  printf("%" PRIx64 " * %" PRIx32 " => %" PRIx64" / %" PRIx32 " => %" PRIx64 "\n",
         t, to_hz, t * to_hz, from_hz, t * to_hz / from_hz);
  u64_t r64 = z_tmcvt(t, from_hz, to_hz, true, false, false, false);
  u64_t r32 = z_tmcvt(t, from_hz, to_hz, true, true, false, false);
  printf("!r32 %" PRIx64 " %" PRIx32 "\n", r64, (u32_t)r64);
  printf("r32 %" PRIx64 " %" PRIx32 "\n", r32, (u32_t)r32);
```
and got
```
lilith[1588]$ gcc -std=c11 /tmp/tm.c && ./a.out
80000000 * 7 => 380000000 / d => 44ec4ec4
!r32 44ec4ec4 44ec4ec4
r32 9d89d89 9d89d89
```
Even though the final result in full precision fits in 32 bits, when called with the flag for a 32-bit result the returned value does not equal a truncated full-precision result.

I'm pretty sure it's wrong in the `div_ratio` branch too, because discarded bits at/above 32 would have affected the result below bit 32 when dividing by a value greater than 1.

I *think* it's ok in the `mul_ratio` branch because the effective operation is multiple, and the high bits aren't affected.

In short, this should have a test dedicated to it.

Reply:
That's actually following the current implementation, though you're right that in the new world it's more likely that these will be used to up-convert to a faster rate (e.g. ticks are now faster than ms by default) and overflow, where before all the conversions were going the other way.  The docs are ambiguous though, promising to return a value in 32 bits of precision sort of implies there will be no overflow.  @nordic-krch was mentioning the same thing.

I don't really know the best path here.  One option would be to do the up-conversion math in 64 bit precision always (at the cost of performance).   Another would be to put an __ASSERT() in there for softer handling.

Reply:
My strong preference is to remove the `u32_t` cast on `t` for all `result32` cases except `mul_ratio`, so the numerator of the division is always 64-bits.  Maybe add an explicit truncation cast of the result to `u32_t` so the compiler can take advantage of any relevant optimizations.

I also think it's still fair to require a coverage test, with values that explicitly use bits 32 and higher in intermediate results.  There are a lot of branches in that code.

I don't know what docs are involved, but asking for a 32 bit result shouldn't mean I can tolerate an incorrect 32-bit result.  Even if `t` is 32 bits and the result is 32 bits, the non-ratio branch requires 64 bits for the multiply in the numerator.

Reply:
The coverage test is coming.  100% code coverage is a project requirement as of this release.

So, if I take your preference correctly: you'd prefer that operations when !CONFIG_SYS_TIMEOUT_64BIT (a 64 bit k_ticks_t makes this all moot, of course) produce full precision results in all cases, even where they don't right now in the current tree? And detecting the overflow as an __ASSERT condition isn't sufficient for you?   I'm not opposed, though there is a performance cost.

Reply:
That sounds about right.  It's like Nordic clocks with TICKS_PER_SEC with odd factors: if you run with a configuration that can't be optimized (i.e. doesn't go through a ratio branch), you're going to pay.

I'm still not clear on where the current implementation that this replicates is.  Maybe if I can see it and its history something else might make sense, but we *know* `t` values passed into this are going to exceed 32 bits, and I just can't see the point of code that efficiently gives you incorrect results.